curvefs/src/metaserver/copyset/concurrent_apply_queue.h

```diff
@@ -0,0 +1,147 @@
+/*
+ *  Copyright (c) 2023 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * File Created: 20230521
+ * Author: Xinlong-Chen
+ */
+
+#ifndef CURVEFS_SRC_METASERVER_COPYSET_CONCURRENT_APPLY_QUEUE_H_
+#define CURVEFS_SRC_METASERVER_COPYSET_CONCURRENT_APPLY_QUEUE_H_
+
+#include <bthread/condition_variable.h>
+#include <bthread/mutex.h>
+#include <glog/logging.h>
+
+#include <atomic>
+#include <condition_variable>  // NOLINT
+#include <mutex>               // NOLINT
+#include <thread>              // NOLINT
+#include <unordered_map>
+#include <utility>
+
+#include "include/curve_compiler_specific.h"
+#include "src/common/concurrent/count_down_event.h"
+#include "src/common/concurrent/task_queue.h"
+#include "curvefs/src/metaserver/copyset/operator_type.h"
+
+using curve::common::CountDownEvent;
+
+namespace curvefs {
+namespace metaserver {
+namespace copyset {
+
+using ::curve::common::GenericTaskQueue;
+
+struct ApplyOption {
+    int wconcurrentsize = 3;
+    int wqueuedepth = 1;
+    int rconcurrentsize = 1;
+    int rqueuedepth = 1;
+    ApplyOption(int wsize, int wdepth, int rsize, int rdepth) :
+        wconcurrentsize(wsize),
+        wqueuedepth(wdepth),
+        rconcurrentsize(rsize),
+        rqueuedepth(rdepth) {}
+    ApplyOption() {}
+};
+
+enum class ThreadPoolType {READ, WRITE};
+
+class CURVE_CACHELINE_ALIGNMENT ApplyQueue {
```

Comment:
Can we reuse this class from bs? it seems only `Schedule` is different.

Reply:
maybe can make ApplyQueue to be a base class, and then override `Schedule` function.
However, it has changed a lot compared with the current version.